### PR TITLE
nixos/roundcube: security improvements

### PR DESCRIPTION
**Pull request by symphorien**
###### Motivation for this change
* database password is world readable in the store
* php update script is ran as root

###### Things done
* If the database is local, use postgres peer authentication. Otherwise, use a password file.
* Leave database initialisation to `postgresql.ensure*`.
* Leave `/var/lib/roundcube` creation to systemd.
* Run php upgrade script as unpriviledged user.
* initialize the des_key (the installer does it, but nixos bypasses the installer...)
* En passant, fix warning about mime types database

This *should* be backward compatible:
* des_key is only used for cookies apparently. So changing it only logs users out.
* changing the default user: /var/lib/roundcube is chowned by systemd
* database owner does not change, and can still log in with peer auth even if it previously used a password

The nixos test still passes, with no modification.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @Ma27 @globin @Vskilet

### CODE REVIEW

**Review by aanderse — COMMENTED** [review]
Fantastic work! Thanks for doing this.

I wonder if @Ma27 has an existing `roundcube` instance that could be used for testing... :thinking:

File: nixos/modules/services/mail/roundcube.nix [hunk 1]
```diff
@@ -116,12 +133,26 @@ in
...
+      ensureDatabases = [ cfg.database.dbname ];
+      ensureUsers = [ {
+        name = cfg.database.username;
+        ensurePermissions = {
+          "DATABASE ${cfg.database.username}" = "ALL PRIVILEGES";
+        };
+      } ];
     };
 
+    users.users.${user} = mkIf localDB {
```
[the remark this answers is not in the document]
Comment (symphorien) [reply nixos/modules/services/mail/roundcube.nix]: the reason is just that it was nginx before, so as to minimize the risk of breakage.

File: nixos/modules/services/mail/roundcube.nix [hunk 2]
```diff
@@ -5,6 +5,8 @@ with lib;
 let
   cfg = config.services.roundcube;
   fpm = config.services.phpfpm.pools.roundcube;
+  localDB = cfg.database.host == "localhost";
```
[the remark this answers is not in the document]
Comment (symphorien) [reply nixos/modules/services/mail/roundcube.nix]: if you want localhost but create yourself I guess you can use 127.0.0.1...

**Comment by aanderse** [comment]
Thanks @symphorien! Let's see what @Ma27 has to say and go from there. Looking over the code you get the :+1: from me, though. Great work!

**Comment by Vskilet** [comment]
Hoo thanks @symphorien ! I'll test now !

**Review by Vskilet — APPROVED** [review]
I tested my old config, the new one with the passwordfile and the peer authentication : it all works !
Great work 👍

**Comment by Ma27** [comment]
> I wonder if @Ma27 has an existing roundcube instance that could be used for testing...

I do :)
If I have sufficient time, I can test this tomorrow :)

File: nixos/modules/services/mail/roundcube.nix [hunk 3]
```diff
@@ -44,7 +46,9 @@ in
       username = mkOption {
         type = types.str;
         default = "roundcube";
-        description = "Username for the postgresql connection";
+        description = "Username for the postgresql connection.
```
Comment (Ma27) [inline nixos/modules/services/mail/roundcube.nix]: Please use multiline strings here.

**Review by Ma27 — CHANGES_REQUESTED** [review]
Thanks a lot for improving this module! Unfortunately I fonud some issues that should be addressed first :) 

> des_key is only used for cookies apparently. So changing it only logs users out.

Not necessarily: I observed on my personal instance that I didn't get logged out, but I couldn't connect to the mailservers (and had to hit "Logout" manually).

**Comment by symphorien** [comment]
Thanks for your review. I fixed the syntax errors you found.

I don't know what to do about the `des_key` problem. Any idea?

**Comment by symphorien** [comment]
On freenode, I was advised to empty the table called `session`. This is implemented in the current version of the PR. Would you mind testing another time, @Ma27 ?
Users are logged out when `/var/lib/roundcube/des_key` is created, so you should remove this file beforehand.

**Review by Ma27 — APPROVED** [review]
Deployed to my roundcube instance and confirmed that truncating the session table if no des-key exists works fine :+1:

The changes seem fine to me and since there are two more :+1:, this should be good to go, so thanks a lot!

**Pull request merged**